Mathlib/MeasureTheory/Group/ModularCharacter.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2024 Noam Atar. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Noam Atar
+-/
+import Mathlib.MeasureTheory.Constructions.Prod.Integral
+import Mathlib.MeasureTheory.Function.LocallyIntegrable
+import Mathlib.MeasureTheory.Group.Integral
+import Mathlib.MeasureTheory.Group.Measure
+import Mathlib.Topology.Metrizable.Urysohn
+import Mathlib.MeasureTheory.Measure.Haar.Unique
+
+
+/-!
+# Modular character of a locally compact group
+
+On a locally compact group, there is a natural character G → ℝ*, which for g : G gives the value
+μ (· * g) / μ, where μ is an (inner regular) Haar measure. This file defines this character, called
+the modular character, and shows that it is multiplicative and independent of the chosen Haar
+measure.
+TODO: Show that the character is continuous.
+
+## Main Declarations
+
+* `modularCharacter_fun`: Given a measure μ, define the modular character associated with this
+measure, using the (noncomputable) `haarScalarFactor`. The result that this does not depend on the
+measure chosen (as long as it is inner regular) is `modularCharacter_fun_eq`.
+* `modularCharacter`: The homomorphism G →* ℝ≥0 whose toFun is `modularCharacter_fun`.
+-/
+
+open MeasureTheory
+open scoped NNReal
+
+namespace MeasureTheory
+
+namespace Measure
+
+variable {G : Type*} [TopologicalSpace G] [Group G] [TopologicalGroup G]
+  [MeasurableSpace G] [BorelSpace G] [LocallyCompactSpace G]
+
+open Classical
+
+/-- Definition of the modularCharacter map -/
+@[to_additive addModularCharacter_fun "Definition of the addModularCharacter map"]
+noncomputable def modularCharacter_fun (μ : Measure G) [IsHaarMeasure μ] : G → ℝ≥0 :=
+  fun g => haarScalarFactor (map (· * g) μ) μ
```

Comment:
I wonder if it would be better to remove the dependence on `μ` by defining the map using the arbitrary choice provided by `MeasureTheory.Measure.haar`, and then show that `map_right_mul_eq_modularCharacter_smul` holds for any Haar measure `μ`. But I'm not sure how well this would work in the situation when `G` isn't second-countable and the uniqueness results are weaker – if the answer is "it's a horrible mess" then feel free to ignore my remark!

Reply:
I'm not sure. @sgouezel, I know you've done work on uniqueness of Haar measures, what do you think?

Reply:
I think this quantity should not depend on mu, so indeed one should probably remove mu from the definition, use an arbitrary choice of Haar measure, and then prove that it works for any Haar measure.

Reply:
So should all of this be under the assumption that G is second countable?

Reply:
No, I don't think so, a locally compact group should be enough. A lemma like integral_isMulLeftInvariant_eq_smul_of_hasCompactSupport or haarScalarFactor_eq_integral_div should be enough to characterize everything, if I'm not mistaken.